pkg/controller/jobs/kubeflow/trainjob/trainjob_controller.go

```diff
@@ -0,0 +1,259 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trainjob
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	kftrainerv2api "github.com/kubeflow/trainer/v2/pkg/apis/trainer/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	jobsetapi "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+	kJobset "sigs.k8s.io/kueue/pkg/controller/jobs/jobset"
+	"sigs.k8s.io/kueue/pkg/features"
+	"sigs.k8s.io/kueue/pkg/podset"
+)
+
+var (
+	gvk                    = kftrainerv2api.GroupVersion.WithKind("TrainJob")
+	FrameworkName          = "trainer.kubeflow.org/trainjob"
+	TrainJobControllerName = "trainer.kubeflow.org/trainjob-controller"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:      SetupIndexes,
+		NewJob:            NewJob,
+		NewReconciler:     NewReconciler,
+		SetupWebhook:      SetupTrainJobWebhook,
+		JobType:           &kftrainerv2api.TrainJob{},
+		AddToScheme:       kftrainerv2api.AddToScheme,
+		MultiKueueAdapter: &multiKueueAdapter{},
+	}))
+}
+
+// +kubebuilder:rbac:groups=scheduling.k8s.io,resources=priorityclasses,verbs=list;get;watch
```

Comment:
Can we audit this and make sure we actually need all these?

There was a thread going on in k8s slack about finalizers and we saw that we added it for Job and then it was copied everywhere else.

Reply:
You are right, most of these are already defined somewhere else so no need to define them here. I've reviewed them all and removed those `rbac` and `verbs` that are not necessary. Please let me know if I missed something